From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Tue, 22 Sep 2020 01:49:19 -0700
Subject: [PATCH] Optimise non-flush packet sending

Original license: GPL-3.0-only
Original project: https://github.com/PaperMC/Paper
Paper pull request: https://github.com/PaperMC/Paper/pull/10172

Places like entity tracking make heavy use of packet sending,
and internally netty will use some very expensive thread wakeup
calls when scheduling.

Thanks to various hacks in ProtocolLib as well as other
plugins, we cannot simply use a queue of packets to group
send on execute. We have to call execute for each packet.

Tux's suggestion here is exactly what was needed - tag
the Runnable indicating it should not make a wakeup call.

Big thanks to Tux for making this possible as I had given
up on this optimisation before he came along.

Locally this patch drops the entity tracker tick by a full 1.5x.

Co-authored-by: Quang Tran <3d7777456@gmail.com>

diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index bc9e45e17ce2ff1cf0b63c5540f8004e7ab0abc3..6ccc4d2a9acce111fbcdaaf3590c8b09b31b2c1d 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -126,6 +126,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public boolean isPending = true;
     public boolean queueImmunity;
     // Paper end - Optimize network
+    private io.netty.channel.SingleThreadEventLoop eventLoop; // Paper - optimise packets that are not flushed
 
     public Connection(final PacketFlow receiving) {
         this.receiving = receiving;
@@ -135,6 +136,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     public void channelActive(final ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
         this.channel = ctx.channel();
+        this.eventLoop = (io.netty.channel.SingleThreadEventLoop) this.channel.eventLoop(); // Paper - optimise packets that are not flushed
         this.address = this.channel.remoteAddress();
         this.preparing = false; // Spigot
         if (this.delayedDisconnect != null) {
@@ -498,6 +500,11 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.channel.eventLoop().inEventLoop()) {
             this.doSendPacket(packet, listener, flush);
         } else {
+            // Paper start - optimise packets that are not flushed
+            if (!flush && org.dreeam.leaf.config.modules.network.OptimizeNonFlushPacketSending.enabled) {
+                this.eventLoop.lazyExecute(() -> this.doSendPacket(packet, listener, false));
+            } else
+                // Paper end - optimise packets that are not flushed
             this.channel.eventLoop().execute(() -> this.doSendPacket(packet, listener, flush));
         }
     }
